File: composition.ts
import { MathFuncCompose, MathFuncOneParam } from "./types";


const add5:MathFuncOneParam = (x) => x + 5
const double:MathFuncOneParam = (x) => x * 2

const compose:MathFuncCompose = (f1, f2) => (x) => f2(f1(x))

const app = compose(add5, double);
console.log(app(3))

File: composition.dart
typedef MathFunctionOneParam = int Function(int);
typedef MathFunctionOneParamCompose = MathFunctionOneParam Function(
    MathFunctionOneParam, MathFunctionOneParam);

MathFunctionOneParam add5 = (x) => x + 5;
MathFunctionOneParam double = (x) => x * 2;

MathFunctionOneParamCompose compose = (fn1, fn2) {
  return (int x) => fn2(fn1(x));
};

void main(List<String> args) {
  final app = compose(add5, double);
  print(app(3));
}
